### DISCUSSION TITLE
Pagination of methods like `GetTriggerKeys`

### DISCUSSION THREAD

**Discussion by drusellers** [quartznet/quartznet#1957] [Q&A]
I'm building out an internal API to expose Quartz state so that I can check on the health of jobs etc. While building this out I noticed that the methods: `GetTriggerKeys`, `GetJobKeys` don't have pagination support. Given that Quartz.net would need to support the various backing data stores I don't suspect this is the easiest thing in the world to add, but I would be interested in providing this PR.

A: Would this be welcome?
B: Anything I should watch out for?
C: Any discussion needed, that I might be missing?

**Comment by lahma** [comment]
There has been some discussion about the feature [here](https://github.com/quartznet/quartznet/issues/1273) . 

> A: Would this be welcome?

I think so. You would probably want this to 3.x release train to enjoy if a bit fast I presume. Basically would need side-by-side / backwards compatible API that would also allow paging.

> B: Anything I should watch out for?

Probably just database differences and StdAdoDelegate would have to resort to (select *).Skip(pagenum * pageSize).Take(pageSize) kind of surprising behavior that would download the database anyway... There are matchers already so they are part of this query interface.

> C: Any discussion needed, that I might be missing?

The above link has some discussion.

**Comment by jafin** [comment]
+1 on paging!
I'd love if it could exposed a `IQuerable<T>` . But I appreciate without something like EF, that is possibly a large amount of work translating a IQueryable to AST/SQL/Other backing stores.  At which point your doing as much work as EF.

Perhaps one option could be,  Expose interfaces with IQuerable/desired methods,  base Quartz provider (existing) could return NotImplemented initially.  But then that opens the door for another lib to provide a Datastore plugin with something like EF that can expose then.  Thus keeping the base Quartz lib EF free.   At some point should someone PR a IQuerable for Quartz base then obviously that can take away the NotImplemented?

This probably aligns more with the Refactor persistence API suggestion in https://github.com/quartznet/quartznet/issues/988
